plugin/completion.py

```diff
@@ -243,6 +243,24 @@ def format_completion(self, item: dict) -> 'Tuple[str, str]':
                 hint = completion_item_kind_names[kind]
         # label is an alternative for insertText if insertText not provided
         insert_text = item.get("insertText") or label
+        # try to handle textEdit if present
+        text_edit = item.get("textEdit")
+        if text_edit:
+            edit_range, edit_text = text_edit.get("range"), text_edit.get("newText")
+            if edit_range and edit_text:
+                range_start, range_end = edit_range.get("start"), edit_range.get("end")
+                if range_start and range_end:
+                    start_row, start_col = range_start.get("line"), range_start.get("character")
+                    end_row, end_col = range_end.get("line"), range_end.get("character")
+                    if all([start_row is not None, start_col is not None,
+                            end_row is not None, end_col is not None]):
+                        last_start = self.last_location - len(self.last_prefix)
+                        last_row, last_col = self.view.rowcol(last_start)
+                        if last_row == start_row == end_row and start_col <= last_col:
+                            # sublime does not support explicit replacement with completion
+                            # at given range, but we try to trim the textEdit range and text
+                            # to the start location of the completion
```

Comment:
Have you tried using the `LspApplyDocumentEditCommand`? (it's in `edit.py`)

Reply:
The problem is that `insert_text` is prepared for each completion before the choice is made by the user.

After choice is made Sublime simply replaces prefix with whatever was the completion text returned by plugin in previous step, at this point there is no way to run plugin callback such as `LspApplyDocumentEditCommand`.

On the other hand language server does not know what is the prefix Sublime tries to complete, it does receive only the location you want to complete and knows document content, hence when you do `$tru`<kbd>Ctrl</kbd><kbd>Space</kbd> language server does not know if the prefix was `tru` or `$tru`.

In my Powershell example Sublime prefix is `tru` however language server is returning `$true`, and after completion we end up with `$$true` (double dollar sign) instead `$true`.

My PR uses extra information from `text_edit` response (if present) to make sure the completion starts exactly where the prefix begins, so Sublime does the proper replacement.

It is not perfect solution, it does assumption that everything happens on single line, and may not work as intended in extreme cases e.g. if language server returns `%true` for `$tru`<kbd>Ctrl</kbd><kbd>Space</kbd>, even it does not make sense, it will be still completed to `$true` instead of `%true`, but we can't do anything about this with the way Sublime currently handles completions.

Reply:
Good points.